### ISSUE TITLE
Conducting experiments on ImageNet using the implementation of FSQ from this repository.

### ISSUE THREAD

**Issue by duchenzhuang** [lucidrains/vector-quantize-pytorch#88]
Many thanks to the authors for implementing the Finite Scalar Quantizer (FSQ). We further conduct experiments on ImageNet, and FSQ proves to be simple and effective. 

This is the repo: https://github.com/duchenzhuang/FSQ-pytorch

**Comment by duchenzhuang** [comment]
> @duchenzhuang nice! what would be helpful is to also try running it against LFQ. i'm curious which one comes out on top

In my current experimental setting, with all other configurations such as network architecture held constant, I observe that switching the quantizer to LFQ didn't work well. Even during the training of the auto-encoder, the reconstruction loss on the training set remained at a relatively high level and did not decrease. 

Additionally, I found a contradiction between the papers on FSQ and LFQ. The FSQ paper suggests that it is advisable for each dimension to have values greater than or equal to 5 (Figure 10), whereas LFQ essentially allows each dimension to have only two values, either +1 or -1.

**Comment by lucidrains** [comment]
thank you Chenzhuang!

disappointing to hear but good to know

**Comment by lucidrains** [comment]
@Lijun-Yu would you like to do one more code review of LFQ, just to make sure my implementation of your work is fair

**Comment by Lijun-Yu** [comment]
For sure, at a glance I see an issue in calculating the `avg_prob` and in turn `codebook_entropy`. I'll do a full review and potentially send a pr to fix shortly

**Comment by lucidrains** [comment]
@Lijun-Yu thank you! 🙏

**Comment by duchenzhuang** [comment]
> For sure, at a glance I see an issue in calculating the `avg_prob` and in turn `codebook_entropy`. I'll do a full review and potentially send a pr to fix shortly

I'm eagerly looking forward to your comprehensive review of LFQ's implementation. Following that, I plan to rerun experiments on my code to validate the Lookup Free Quantizer.

**Comment by lucidrains** [comment]
@Lijun-Yu could you possibly put the issue you observe into writing? probably faster that i just put in the patch

**Comment by Lijun-Yu** [comment]
Sorry for the delay. I've started running verification with the same auto encoder architecture following your examples. And I found more issues than I expected - the current implementation is giving 1% utilization with LFQ. Let me investigate more and update here in a few hours.

**Comment by Lijun-Yu** [comment]
> In addition, I've noticed that when I set the auto-encoder's downsampling factor to 16 (e.g., compressing a 256 * 256 image into a 16 * 16 latent space), the reconstruction quality is significantly poorer. 
> 
> 
> 
> I've observed that both VQGAN and MagViTv2 use a downsampling factor of 16, but even though VQGAN has open-sourced their code, it seems challenging to reproduce results similar to their paper. 
> 
> 
> 
> Do you have any recommendations or references to address this issue? @Lijun-Yu @lucidrains 

256x256 -> 16x16 is definitely doable as have been done in VQGAN and MaskGIT. And MAGVITv2 also show that we can do 512x512 -> 16x16 with SOTA generation quality. However, such high compression rate puts higher pressure on the model - in general you need more blocks/layers to do more down/upsampling and also to build a deeper/larger model. You may find Fig 7 from arxiv.org/pdf/2212.05199 helpful if you change all 3D conv to 2D and adapt it for 256x256

**Comment by duchenzhuang** [comment]
> > In addition, I've noticed that when I set the auto-encoder's downsampling factor to 16 (e.g., compressing a 256 * 256 image into a 16 * 16 latent space), the reconstruction quality is significantly poorer.
> > I've observed that both VQGAN and MagViTv2 use a downsampling factor of 16, but even though VQGAN has open-sourced their code, it seems challenging to reproduce results similar to their paper.
> > Do you have any recommendations or references to address this issue? @Lijun-Yu @lucidrains
> 
> 256x256 -> 16x16 is definitely doable as have been done in VQGAN and MaskGIT. And MAGVITv2 also show that we can do 512x512 -> 16x16 with SOTA generation quality. However, such high compression rate puts higher pressure on the model - in general you need more blocks/layers to do more down/upsampling and also to build a deeper/larger model. You may find Fig 7 from arxiv.org/pdf/2212.05199 helpful if you change all 3D conv to 2D and adapt it for 256x256

Thank you once again for your response. I would like to inquire about the necessity of GAN loss when training a VQVAE with high compression ratios (such as 256 * 256->16 * 16 or 512 * 512->16 * 16). In previous experiments, I did not include GAN loss due to its instability. Therefore, I would like to seek your insights on the importance of GAN loss.

**Comment by Lijun-Yu** [comment]
> > > In addition, I've noticed that when I set the auto-encoder's downsampling factor to 16 (e.g., compressing a 256 * 256 image into a 16 * 16 latent space), the reconstruction quality is significantly poorer.
> > > I've observed that both VQGAN and MagViTv2 use a downsampling factor of 16, but even though VQGAN has open-sourced their code, it seems challenging to reproduce results similar to their paper.
> > > Do you have any recommendations or references to address this issue? @Lijun-Yu @lucidrains
> > 
> > 
> > 256x256 -> 16x16 is definitely doable as have been done in VQGAN and MaskGIT. And MAGVITv2 also show that we can do 512x512 -> 16x16 with SOTA generation quality. However, such high compression rate puts higher pressure on the model - in general you need more blocks/layers to do more down/upsampling and also to build a deeper/larger model. You may find Fig 7 from arxiv.org/pdf/2212.05199 helpful if you change all 3D conv to 2D and adapt it for 256x256
> 
> Thank you once again for your response. I would like to inquire about the necessity of GAN loss when training a VQVAE with high compression ratios (such as 256 * 256->16 * 16 or 512 * 512->16 * 16). In previous experiments, I did not include GAN loss due to its instability. Therefore, I would like to seek your insights on the importance of GAN loss.

In general, GAN loss leads to better visual quality if that's what you care about. And the architecture/capacity of the discriminator matters quite a bit, e.g. StyleGAN would be better/more stable than PatchGAN. In addition, GAN loss has a significant impact on FID/FVD scores if you care about those metrics.

**Comment by Lijun-Yu** [comment]
I fixed a few issues in #89 , @duchenzhuang maybe you can have a try on ImageNet as well
The default parameters in `lookup_free_quantization.py` should be reasonable enough to work. But you may need to train a generation model to fully see its benefits compared to other methods.

**Comment by lucidrains** [comment]
@Lijun-Yu thank you for the PR! very excited to try breeding LFQ latents in a genetic algorithm setting

**Comment by duchenzhuang** [comment]
> I fixed a few issues in #89 , @duchenzhuang maybe you can have a try on ImageNet as well The default parameters in `lookup_free_quantization.py` should be reasonable enough to work. But you may need to train a generation model to fully see its benefits compared to other methods.

Thank you, Lijun. I will try it as soon as possible.

**Comment by duchenzhuang** [comment]
Thank you very much, @Lijun-Yu @lucidrains! The Lookup Free Quantizer works effectively in my experiment! 

I would like to inquire with Lijun @Lijun-Yu  about the selection of hyperparameters for LFQ. 
Is commitment_loss_weight fixed at 0.25 for all experiments? 
I noticed that entropy_loss_weight is defaulting to 0.1 in LFQ, but in your example, the weight for entropy loss is even smaller. Could you please advise on a suitable value for entropy_loss_weight, especially for larger datasets like ImageNet? 
As the Codebook size increases, how should the weights for entropy loss be adjusted? 

Thank you.

**Comment by duchenzhuang** [comment]
By the way, @Lijun-Yu , in your  experiments, did you only use perceptual loss and adversarial loss without employing L2 loss or L1 loss for image reconstruction when training the tokenizer?

**Comment by lucidrains** [comment]
![meme](https://github.com/lucidrains/vector-quantize-pytorch/assets/108653/782e7934-3964-4bda-be02-7780320ef0d4)

**Comment by Lijun-Yu** [comment]
> Thank you very much, @Lijun-Yu @lucidrains! The Lookup Free Quantizer works effectively in my experiment!
> 
> I would like to inquire with Lijun @Lijun-Yu about the selection of hyperparameters for LFQ. Is commitment_loss_weight fixed at 0.25 for all experiments? I noticed that entropy_loss_weight is defaulting to 0.1 in LFQ, but in your example, the weight for entropy loss is even smaller. Could you please advise on a suitable value for entropy_loss_weight, especially for larger datasets like ImageNet? As the Codebook size increases, how should the weights for entropy loss be adjusted?
> 
> Thank you.

In all our experiments in the paper, we use commitment loss weight 0.25 and entropy loss weight 0.1, which has been tested for ImageNet/Kinetics/larger datasets across vocabulary sizes from 2^10 to 2^40. (In the next arxiv revision we'll provide a hyperparameter list to clarify these.)


> By the way, @Lijun-Yu , in your experiments, did you only use perceptual loss and adversarial loss without employing L2 loss or L1 loss for image reconstruction when training the tokenizer?

We use L2 + perceptual + adversarial. You may find many implementation details in the official repo for magvitv1 https://github.com/google-research/magvit/tree/main. Magvitv2 is similar unless specifically noted in our paper.

**Comment by Lijun-Yu** [comment]
> but in your example, the weight for entropy loss is even smaller

The toy example uses much smaller resolution/model/dataset, so it's not representative enough. It's still okay to use 0.1 in this toy setup, but the training would focus too much on optimizing the entropy and lead to somewhat worse reconstruction, which is why I reduced the weight. This is usually not the case when you train deeper model on larger datasets.

In addition, I want to remind you of the entropy annealing schedule which is quite useful for large vocabularies. @duchenzhuang 

<img width="619" alt="image" src="https://github.com/lucidrains/vector-quantize-pytorch/assets/36872895/afa54e48-ebc9-4177-8d7e-d3a188f7f9a0">

**Comment by duchenzhuang** [comment]
> > but in your example, the weight for entropy loss is even smaller
> 
> The toy example uses much smaller resolution/vocabulary/model/dataset, so it's not representative enough. It's still okay to use 0.1 in this toy setup, but the training would focus too much on optimizing the entropy and lead to somewhat worse reconstruction, which is why I reduced the weight. This is usually not the case when you train deeper model on larger datasets with perceptual and adversarial losses.
> 
> In addition, I want to remind you of the entropy annealing schedule which is quite useful for large vocabularies. @duchenzhuang
> 
> <img alt="image" width="619" src="https://private-user-images.githubusercontent.com/36872895/287498958-afa54e48-ebc9-4177-8d7e-d3a188f7f9a0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE1ODQ1OTAsIm5iZiI6MTcwMTU4NDI5MCwicGF0aCI6Ii8zNjg3Mjg5NS8yODc0OTg5NTgtYWZhNTRlNDgtZWJjOS00MTc3LThkN2UtZDNhMTg4ZjdmOWEwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjAzVDA2MTgxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdkZGM3YjI1OWNmNmNjYzM3Yzk1NzIyMzBmMmVhMGNjMDk5NzMxNGY0ODFiZWUzZTMyZTAzNTU3ZDY1MTdjMTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.6j0Ub-cAG8UXm5ixQLSQjzEnNPcbWz_XEt2dbJEU3eQ">

Thank you, Lijun @Lijun-Yu .

**Comment by lucidrains** [comment]
@Lijun-Yu congrats Lijun! https://github.com/lucidrains/meshgpt-pytorch/issues/9#issuecomment-1852588734
